t/01-wasm/directives/008-flag_directive_wasmtime.t

```diff
@@ -87,7 +87,111 @@ qr/setting flag: "debug_info=off"/
 
 
 
-=== TEST 5: flag directive - wasmtime - consume_fuel - on
+=== TEST 5: flag directive - wasmtime - cache_config_load - missing file
+--- skip_eval: 4: $::nginxV !~ m/wasmtime/
+--- main_config
+    wasm {
+        wasmtime {
+            flag cache_config_load missing_file;
```

Comment:
Should we implement this a level above the `flag` directive? It reads a little weird or is it just me? E.g:
```
wasmtime {
    cache_config  /path/to/file;
}
```

Reply:
Does wasmtime expose an API for setting its cache parameters directly? It's not the end of the world if we have to generate the toml file and pass it in, but something like this would be a lot more ergonomic:

```
wasmtime {
  # implies cache.enabled = true
  cache_directory = /path/to/dir;
}
```

Reply:
It does read a bit weird — I originally called it just `cache_config`, but then when I checked our docs to document it, the docs pointed to the Wasmtime Config struct page, and all of our flag names were 1-to-1 with the method calls for that struct, so I aligned it to that method call name for consistency.

The other top-level entries we have tend to be shared across VMs (`backtraces`, `compiler`) and this one is very Wasmtime-specific (it would end up declared at `ngx_wasm_core_module.c`?), so maybe keeping it under `flag` is fine...

As for the name, I wouldn't mind calling it `flag cache_config /path/to/file` and then specifically documenting it in `docs/DIRECTIVES.md`.

Reply:
@flrgh [it doesn't](https://docs.wasmtime.dev/api/wasmtime/struct.Config.html#method.cache_config_load), and the actual .toml has [a lot of options](https://docs.wasmtime.dev/cli-cache.html), so it's probably for the best if you have the flexibility to write them in the .toml on your end rather than relying on which options are exposed or not by wasmx given a particular version of wasmtime...

Reply:
It could still be wasmtime specific but not part of `flag` since we have the `wasmtime{}` block for the purpose of runtime-specific configuration directives. All `flag` handlers are supposed to be generic (e.g. `bool_flag_handler` or `size_flag_handler`) and receive a `wrt_setter` pointer, which is unused in the current implementation of `cache_config_load_flag_handler`. We could have `string_flag_handler` and give the `wasmtime_config_cache_config_load` as the `wrt_setter`, if that works (didn't check).

Reply:
@thibaultcha I attempted to implement it as a `wasmtime`-level `config_cache` directive in this branch: https://github.com/Kong/ngx_wasm_module/tree/feat/wasmtime-config-cache-no-flag

...but it's giving me `invalid number of arguments in "cache_config"` errors, and I can't figure out why. I used the same NGX_CONF_TAKE1 flag as the `"compiler"` directive right below it.

Any idea why?

Reply:
It looks like an issue with our definitions :/ We should hit [this](https://github.com/nginx/nginx/blob/master/src/core/ngx_conf_file.c#L419) expression successfully and pass the arguments check, but we don't pass [these two](https://github.com/nginx/nginx/blob/master/src/core/ngx_conf_file.c#L415-L417) previous conditions (I gdb'ed it).

Reply:
This took care of it:
```patch
diff --git a/src/wasm/ngx_wasm.h b/src/wasm/ngx_wasm.h
index d832f026..c570d10b 100644
--- a/src/wasm/ngx_wasm.h
+++ b/src/wasm/ngx_wasm.h
@@ -14,10 +14,10 @@
 #define NGX_LOG_WASM_NYI             NGX_LOG_ALERT
 
 #define NGX_WASM_MODULE              0x5741534d   /* "WASM" */
-#define NGX_WASM_CONF                0x00300000
-#define NGX_WASMTIME_CONF            0x00001000
-#define NGX_WASMER_CONF              0x00002000
-#define NGX_V8_CONF                  0x00004000
+#define NGX_WASM_CONF                0x10000000
+#define NGX_WASMTIME_CONF            0x20000000
+#define NGX_WASMER_CONF              0x40000000
+#define NGX_V8_CONF                  0x80000000
 
 #define NGX_WASM_DONE_PHASE          15
 #define NGX_WASM_BACKGROUND_PHASE    16
diff --git a/src/wasm/wrt/ngx_wrt_wasmtime.c b/src/wasm/wrt/ngx_wrt_wasmtime.c
index 802ab5ff..2ed0f1c2 100644
--- a/src/wasm/wrt/ngx_wrt_wasmtime.c
+++ b/src/wasm/wrt/ngx_wrt_wasmtime.c
@@ -171,7 +171,7 @@ ngx_wasmtime_init_conf(ngx_wavm_conf_t *conf, ngx_log_t *log)
     if (conf->cache_config.len) {
         ngx_wavm_log_error(NGX_LOG_INFO, log, NULL,
                            "setting wasmtime cache config file: \"%V\"",
-                           conf->cache_config);
+                           &conf->cache_config);
 
         pathname = ngx_calloc(conf->cache_config.len + 1, log);
         if (pathname == NULL) {
```

I hope the `NGX_WASM_CONF` change works for the whole suite.

Reply:
Thanks! Rebased.